crates/red_knot_python_semantic/src/types/builder.rs

```diff
@@ -27,79 +27,87 @@
+            elements: vec![],
         }
     }
 
     /// Adds a type to this union.
     pub(crate) fn add(mut self, ty: Type<'db>) -> Self {
         match ty {
             Type::Union(union) => {
-                for element in union.elements(self.db) {
+                let new_elements = union.elements(self.db);
+                self.elements.reserve(new_elements.len());
+                for element in &**new_elements {
                     self = self.add(*element);
                 }
             }
             Type::Never => {}
             _ => {
-                let mut remove = vec![];
-                for element in &self.elements {
+                let bool_pair = if let Type::BooleanLiteral(b) = ty {
+                    Some(Type::BooleanLiteral(!b))
+                } else {
+                    None
+                };
+                let mut to_add = ty;
+                let mut to_remove = vec![];
+                for (index, element) in self.elements.iter().enumerate() {
+                    if Some(*element) == bool_pair {
+                        to_add = builtins_symbol_ty(self.db, "bool");
+                        to_remove.push(index);
+                        // The type we are adding is a BooleanLiteral, which doesn't have any
+                        // subtypes. And we just found that the union already contained our
+                        // mirror-image BooleanLiteral, so it can't also contain bool or any
+                        // supertype of bool. Therefore, we are done.
+                        break;
+                    }
                     if ty.is_subtype_of(self.db, *element) {
                         return self;
                     } else if element.is_subtype_of(self.db, ty) {
-                        remove.push(*element);
+                        to_remove.push(index);
                     }
                 }
-                for element in remove {
-                    self.elements.remove(&element);
+                match to_remove[..] {
+                    [] => self.elements.push(to_add),
+                    [index] => self.elements[index] = to_add,
+                    _ => {
+                        let mut retain =
+                            Vec::with_capacity(self.elements.len() - to_remove.len() + 1);
+                        let mut remove_iter = to_remove.iter();
+                        let mut elements_iter = self.elements.iter().enumerate();
+                        'outer: for remove_index in remove_iter.by_ref() {
+                            for (index, element) in elements_iter.by_ref() {
+                                if index == *remove_index {
+                                    continue 'outer;
+                                }
+                                retain.push(*element);
+                            }
+                        }
+                        retain.extend(elements_iter.map(|(_, element)| element));
```

Comment:
Thanks for all the good suggestions! Still evaluating both yours and Micha's, but one note here: the initial simplification you suggest in the suggested change here doesn't work, because it causes the enumeration to start over at 0 each time we re-enter the inner loop. But this was also useful, because it made me realize the tests were inadequate.